base/cvd/cuttlefish/host/libs/zip/disjoint_range_set.cc

```diff
@@ -0,0 +1,162 @@
+    return start_ <= other.start_ && end_ >= other.end_;
+  }
+
+  std::optional<Range> MergeIfPossible(const Range& other) const {
+    bool intersects = Contains(other.start_) || Contains(other.end_) ||
+                      other.Contains(start_) || other.Contains(end_);
+    bool adjacent = start_ == other.end_ || end_ == other.start_;
+    if (!intersects && !adjacent) {
+      return std::nullopt;
+    }
+    Range new_range(std::min(start_, other.start_), std::max(end_, other.end_));
+    return new_range;
+  }
+
+  bool operator<(const Range& other) const { return start_ < other.start_; }
+  bool operator==(const Range& other) const {
+    return start_ == other.start_ && end_ == other.end_;
+  }
+
+  std::pair<uint64_t, uint64_t> AsPair() const {
+    return std::make_pair(start_, end_);
+  }
+
+ private:
+  uint64_t start_;
+  uint64_t end_;
+};
+
+using Iterator = std::set<Range>::iterator;
+
+}  // namespace
+
+struct DisjointRangeSet::Impl {
+  std::set<Range> ranges_;
+};
+
+DisjointRangeSet::DisjointRangeSet() : impl_(new Impl) {}
+
+DisjointRangeSet::DisjointRangeSet(const DisjointRangeSet& other)
+    : impl_(new Impl) {
+  impl_->ranges_ = other.impl_->ranges_;
+}
+
+DisjointRangeSet::DisjointRangeSet(DisjointRangeSet&& other) : impl_(new Impl) {
+  std::swap(impl_, other.impl_);
+}
+
+DisjointRangeSet& DisjointRangeSet::operator=(const DisjointRangeSet& other) {
+  impl_->ranges_ = other.impl_->ranges_;
+  return *this;
+}
+
+DisjointRangeSet& DisjointRangeSet::operator=(DisjointRangeSet&& other) {
+  impl_.reset(new Impl);
+  std::swap(impl_, other.impl_);
+  return *this;
+}
+
+DisjointRangeSet::~DisjointRangeSet() = default;
+
+bool DisjointRangeSet::ContainsRange(uint64_t start, uint64_t end) const {
+  const std::set<Range>& ranges = impl_->ranges_;
+
+  if (ranges.empty()) {
+    return false;
+  }
+  Range range(start, end);
+
+  Iterator range_it = ranges.upper_bound(range);
+  if (range_it != ranges.begin()) {
+    range_it--;
+  }
+  return range_it->Contains(range);
+}
+
+void DisjointRangeSet::InsertRange(uint64_t start, uint64_t end) {
+  if (ContainsRange(start, end)) {
+    return;
+  }
+
+  std::set<Range>& ranges = impl_->ranges_;
+  Range new_range(start, end);
+
+  Iterator it = ranges.emplace(new_range).first;
+  if (std::optional<Range> merge = it->MergeIfPossible(new_range); merge) {
+    ranges.erase(it);
+    it = ranges.emplace(new_range).first;
+  }
+
+  while (it != ranges.begin()) {
+    it--;
+    if (std::optional<Range> merge = it->MergeIfPossible(new_range); merge) {
+      ranges.erase(it);
+      it = ranges.emplace(new_range = *merge).first;
+    } else {
+      it++;
+      break;
+    }
+  }
+
+  it++;
+  while (it != ranges.end()) {
+    if (std::optional<Range> merge = it->MergeIfPossible(new_range); merge) {
+      it--;
+      it = ranges.erase(it);
+      ranges.erase(it);
+      it = ranges.emplace(new_range = *merge).first;
+      it++;
+    } else {
+      break;
+    }
+  }
```

Comment:
This would be simpler to implement and easier to read if you remove and merge all relevant ranges before inserting the new (merged) range:
```
auto it = ranges.upper_bound(new_range);
while (it != ranges.end() && new_range.Touches(*it)) {
    new_range = new_range.MergeIfPossible(*it);
    it = ranges.erase(it)
}
if (it != ranges.begin() && new_range.Touches(*(--it))) {
    new_range = new_range.MergeIfPossible(*it);
    it = ranges.erase(it);
}
ranges.emplace(new_range);
```

Reply:
It looks a lot nicer for sure but it did take a while to convince myself this is correct. Changed to use this implementation.